jerry-core/jcontext/jcontext.h

```diff
@@ -121,52 +121,65 @@ typedef struct
 #endif /* JERRY_VALGRIND_FREYA */
 } jerry_context_t;
 
+#ifndef CONFIG_ECMA_LCACHE_DISABLE
+/**
+ * Hash table for caching the last access of properties.
+ */
+typedef struct
+{
+  ecma_lcache_hash_entry_t table[ECMA_LCACHE_HASH_ROWS_COUNT][ECMA_LCACHE_HASH_ROW_LENGTH];
+} jerry_hash_table_t;
+#endif /* !CONFIG_ECMA_LCACHE_DISABLE */
+
+#ifdef JERRY_ENABLE_EXTERNAL_CONTEXT
+
+#ifndef JERRY_GET_CURRENT_INSTANCE
+
+/**
+ * Default function if JERRY_GET_CURRENT_INSTANCE is not defined.
+ */
+#define JERRY_GET_CURRENT_INSTANCE() (jerry_port_get_current_instance ())
+
+#endif /* !JERRY_GET_CURRENT_INSTANCE */
+
 /**
- * Description of jerry instance, which contains the meta data and buffer.
+ * Description of jerry instance which is the header of the context space.
  */
 struct jerry_instance_t
 {
-  uint32_t heap_size; /**< size of the heap */
+  jerry_context_t context; /**< the context of the instance */
 #ifndef JERRY_SYSTEM_ALLOCATOR
-  uint8_t *heap_p; /**< point to the entrance of the heap in buffer */
-#endif /* !JERRY_SYSTEM_ALLOCATOR */
+  uint32_t heap_size; /**< size of the heap */
+#endif
 #ifndef CONFIG_ECMA_LCACHE_DISABLE
   uint8_t *lcache_p; /**< point to the entrance of the lcache in buffer */
 #endif /* !CONFIG_ECMA_LCACHE_DISABLE */
-  jerry_context_t context; /**< the context of the instance */
-  uint8_t buffer[]; /**< the flexible array for storing context, heap and lcache */
 };
 
-#ifndef CONFIG_ECMA_LCACHE_DISABLE
 /**
- * Hash table for caching the last access of properties.
+ * Contains the JMEM aligned size of jerry_instance_t.
  */
-typedef struct
-{
-  ecma_lcache_hash_entry_t table[ECMA_LCACHE_HASH_ROWS_COUNT][ECMA_LCACHE_HASH_ROW_LENGTH];
-} jerry_hash_table_t;
-#endif /* !CONFIG_ECMA_LCACHE_DISABLE */
-
-#ifdef JERRY_ENABLE_EXTERNAL_CONTEXT
+#define ALIGNED_SIZE_OF_JERRY_INSTANCE \
+  (JERRY_ALIGNUP (sizeof (struct jerry_instance_t), JMEM_ALIGNMENT))
 
 /**
  * This part is for Jerry which enable external context.
  */
-
 typedef struct
 {
   jmem_heap_free_t first; /**< first node in free region list */
   uint8_t area[]; /**< heap area */
 } jmem_heap_t;
 
-#define JERRY_CONTEXT(field) (jerry_port_get_current_instance ()->context.field)
+#define JERRY_CONTEXT(field) (JERRY_GET_CURRENT_INSTANCE ()->context.field)
 
 #ifndef JERRY_SYSTEM_ALLOCATOR
 
 static inline jmem_heap_t * __attr_always_inline___
 jerry_context_get_current_heap (void)
 {
-  return (jmem_heap_t *) (jerry_port_get_current_instance ()->heap_p);
+  uint8_t *instance_p = (uint8_t *) JERRY_GET_CURRENT_INSTANCE ();
+  return (jmem_heap_t *) (instance_p + ALIGNED_SIZE_OF_JERRY_INSTANCE);
```

Comment:
I have been thinking about it. The easiest way to fix it is:

a) passing an alignment requirement to the malloc callback
b) align the returned instance_p and add a function which tells the padding
c) keep the current one, where have to do an dereference when accessing the heap

Probably a is the easiest, since we could simply use posix_memalign. Do you have a better idea?